### ISSUE TITLE
Rewrite Imminent

### ISSUE THREAD

**Issue by uncomfyhalomacro** [openSUSE-Rust/obs-service-cargo#108]
tldr; Apply what we learned from https://github.com/openSUSE-Rust/obs-service-cargo-vendor-home-registry

# Rationale

I have to rewrite the logic regarding `obs-service-cargo-vendor` and it might improve some things on how we vendor.

There are things I found out while I was writing https://github.com/openSUSE-Rust/obs-service-cargo-vendor-home-registry. And I think the one in there makes sense regarding how we manage deps. And I think I have a good feeling about that if we copy/borrow/steal over most code from https://github.com/openSUSE-Rust/obs-service-cargo-vendor-home-registry, a lot of the inconsistencies will be mitigated, if not removed.

## Things I learned
- A lot of code from vendor-home-registry dawned on me that it was the right approach on how we vendor as well using `cargo vendor` (although, there it uses `cargo fetch` which for me makes vendoring easier to write).
- We should always pass `--locked` flag on `cargo vendor` and `cargo generate-lockfiles`. If it fails, that means the user should change the `update` setting to `true`.
- The introduction of `versioned-dirs` by default introduces performance improvements during vendor because it says so in the help section of `cargo vendor`:
   >  Normally versions are only added to disambiguate multiple versions of the same package. This option causes all directories in the “vendor” directory to be versioned, which makes it easier to track the history of vendored packages over time, and can help with the performance of re-vendoring when only a subset of the packages have changed.
- The `no-root-manifest` flag with a combination of `custom-root` flag from vendor-home-registry is a good idea to have here as well.

cc @Firstyear @Pi-Cla (for cheers and opinions) :P

Related issues: #97 #100 #96

**Comment by uncomfyhalomacro** [comment]
# From OBS Cargo Vendor Home Registry

### Setting Updates

**Update**: False

obs-cargo-vendor-home-registry
![image](https://github.com/user-attachments/assets/90c63e2f-7b0e-498d-ab6d-7ef018ae713c)

python-tokenizers
![image](https://github.com/user-attachments/assets/29ccd1cd-adea-459d-8a4c-182927454db8)

**Update**: True

obs-cargo-vendor-home-registry
![image](https://github.com/user-attachments/assets/4d3c33af-df32-4488-b7c9-497ac438562e)

python-tokenizers
![image](https://github.com/user-attachments/assets/3b2299f6-4a5a-49e9-8b8d-00a59c1b881a)

**Comment by uncomfyhalomacro** [comment]
Although, a removed lockfile or a non-existent lockfile might need some improvements on how we convey things to the users.

![image](https://github.com/user-attachments/assets/b10824c2-b3e3-41b1-b6c3-5d166d671559)

**Comment by Firstyear** [comment]
Combination makes long term maintenance easier because you only have one place to change and package to release. But it also means if you "rewrite" you have a bigger surface area to alter as you go.

I think the only advice I have is not to alter the "user facing" side. I think this should be contained to internals mostly.

**Comment by uncomfyhalomacro** [comment]
> Combination makes long term maintenance easier because you only have one place to change and package to release. But it also means if you "rewrite" you have a bigger surface area to alter as you go.
> 
> I think the only advice I have is not to alter the "user facing" side. I think this should be contained to internals mostly.

Okay so there should be an option to use either vendor or registry but not both.
